二进制文件 batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/BCM4345C0.hcd 和 batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/BCM4345C0.hcd 不同
二进制文件 batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.bin 和 batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.bin 不同
二进制文件 batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.rockchip,tn3399_v3.bin 和 batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.rockchip,tn3399_v3.bin 不同
diff -uprN batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.rockchip,tn3399_v3.txt batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.rockchip,tn3399_v3.txt
--- batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.rockchip,tn3399_v3.txt	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.rockchip,tn3399_v3.txt	2022-01-08 17:50:06.000000000 +0800
@@ -0,0 +1,83 @@
+# SPDX-License-Identifier: GPL-2.0+
+# (C) Copyright 2018 Raspberry Pi (Trading) Ltd.
+# NVRAM config file for the BCM43455 WiFi/BT chip as found on the
+# Raspberry Pi 4 Model B
+aa2g=1
+aa5g=1
+AvVmid_c0=1,165,2,100,2,100,2,100,2,100
+boardflags=0x00480201
+boardflags2=0x40800000
+boardflags3=0x44200100
+boardrev=0x1304
+boardtype=0x6e4
+btc_mode=1
+btc_params1=0x7530
+btc_params8=0x4e20
+btc_params50=0x972c
+cbfilttype=1
+cckPwrIdxCorr=3
+cckTssiDelay=150
+deadman_to=481500000
+devid=0x43ab
+dot11agofdmhrbw202gpo=0x4442
+dot11b_opts=0x3aa85
+ed_thresh2g=-54
+ed_thresh5g=-54
+eu_edthresh2g=-54
+eu_edthresh5g=-54
+extpagain2g=2
+extpagain5g=2
+fdsslevel_ch11=6
+femctrl=0
+itrsw=1
+ldo1=4
+ltecxfnsel=0x22
+ltecxgcigpio=0x32
+ltecxmux=0
+ltecxpadnum=0x0504
+macaddr=b8:27:eb:74:f2:6c
+manfid=0x2d0
+maxp2ga0=80
+maxp5ga0=82,82,82,82
+mcsbw202gpo=0x98444422
+mcsbw205ghpo=0xb9555000
+mcsbw205glpo=0xb9555000
+mcsbw205gmpo=0xb9555000
+mcsbw402gpo=0x98444422
+mcsbw405ghpo=0xb9555000
+mcsbw405glpo=0xb9555000
+mcsbw405gmpo=0xb9555000
+mcsbw805ghpo=0xb9555000
+mcsbw805glpo=0xb9555000
+mcsbw805gmpo=0xb9555000
+nocrc=1
+ofdmlrbw202gpo=0x0022
+ofdmTssiDelay=150
+pa2ga0=-129,6525,-718
+pa2ga1=-149,4408,-601
+pa5ga0=-185,6836,-815,-186,6838,-815,-184,6859,-815,-184,6882,-818
+pa5ga1=-202,4285,-574,-201,4312,-578,-196,4391,-586,-201,4294,-575
+pdoffset2g40ma0=16
+pdoffset2gperchan=0,-2,1,0,1,0,1,1,1,0,0,-1,-1,0
+pdoffset40ma0=0x8888
+pdoffset80ma0=0x8888
+pdoffsetcckma0=2
+phycal_tempdelta=15
+prodid=0x06e4
+rawtempsense=0x1ff
+rxchain=1
+sromrev=11
+swctrlmap_2g=0x00000000,0x00000000,0x00000000,0x010000,0x3ff
+swctrlmap_5g=0x00100010,0x00200020,0x00200020,0x010000,0x3fe
+swctrlmapext_2g=0x00000000,0x00000000,0x00000000,0x000000,0x3
+swctrlmapext_5g=0x00000000,0x00000000,0x00000000,0x000000,0x3
+tssipos2g=1
+tssipos5g=1
+tworangetssi2g=1
+tworangetssi5g=1
+txchain=1
+txpwr2gAdcScale=1
+txpwr5gAdcScale=1
+vcodivmode=1
+vendid=0x14e4
+xtalfreq=37400
diff -uprN batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.txt batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.txt
--- batocera.linux-b34_origin/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.txt	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/board/batocera/rockchip/rk3399/fsoverlay/lib/firmware/brcm/brcmfmac43455-sdio.txt	2022-01-08 17:50:06.000000000 +0800
@@ -0,0 +1,83 @@
+# SPDX-License-Identifier: GPL-2.0+
+# (C) Copyright 2018 Raspberry Pi (Trading) Ltd.
+# NVRAM config file for the BCM43455 WiFi/BT chip as found on the
+# Raspberry Pi 4 Model B
+aa2g=1
+aa5g=1
+AvVmid_c0=1,165,2,100,2,100,2,100,2,100
+boardflags=0x00480201
+boardflags2=0x40800000
+boardflags3=0x44200100
+boardrev=0x1304
+boardtype=0x6e4
+btc_mode=1
+btc_params1=0x7530
+btc_params8=0x4e20
+btc_params50=0x972c
+cbfilttype=1
+cckPwrIdxCorr=3
+cckTssiDelay=150
+deadman_to=481500000
+devid=0x43ab
+dot11agofdmhrbw202gpo=0x4442
+dot11b_opts=0x3aa85
+ed_thresh2g=-54
+ed_thresh5g=-54
+eu_edthresh2g=-54
+eu_edthresh5g=-54
+extpagain2g=2
+extpagain5g=2
+fdsslevel_ch11=6
+femctrl=0
+itrsw=1
+ldo1=4
+ltecxfnsel=0x22
+ltecxgcigpio=0x32
+ltecxmux=0
+ltecxpadnum=0x0504
+macaddr=b8:27:eb:74:f2:6c
+manfid=0x2d0
+maxp2ga0=80
+maxp5ga0=82,82,82,82
+mcsbw202gpo=0x98444422
+mcsbw205ghpo=0xb9555000
+mcsbw205glpo=0xb9555000
+mcsbw205gmpo=0xb9555000
+mcsbw402gpo=0x98444422
+mcsbw405ghpo=0xb9555000
+mcsbw405glpo=0xb9555000
+mcsbw405gmpo=0xb9555000
+mcsbw805ghpo=0xb9555000
+mcsbw805glpo=0xb9555000
+mcsbw805gmpo=0xb9555000
+nocrc=1
+ofdmlrbw202gpo=0x0022
+ofdmTssiDelay=150
+pa2ga0=-129,6525,-718
+pa2ga1=-149,4408,-601
+pa5ga0=-185,6836,-815,-186,6838,-815,-184,6859,-815,-184,6882,-818
+pa5ga1=-202,4285,-574,-201,4312,-578,-196,4391,-586,-201,4294,-575
+pdoffset2g40ma0=16
+pdoffset2gperchan=0,-2,1,0,1,0,1,1,1,0,0,-1,-1,0
+pdoffset40ma0=0x8888
+pdoffset80ma0=0x8888
+pdoffsetcckma0=2
+phycal_tempdelta=15
+prodid=0x06e4
+rawtempsense=0x1ff
+rxchain=1
+sromrev=11
+swctrlmap_2g=0x00000000,0x00000000,0x00000000,0x010000,0x3ff
+swctrlmap_5g=0x00100010,0x00200020,0x00200020,0x010000,0x3fe
+swctrlmapext_2g=0x00000000,0x00000000,0x00000000,0x000000,0x3
+swctrlmapext_5g=0x00000000,0x00000000,0x00000000,0x000000,0x3
+tssipos2g=1
+tssipos5g=1
+tworangetssi2g=1
+tworangetssi5g=1
+txchain=1
+txpwr2gAdcScale=1
+txpwr5gAdcScale=1
+vcodivmode=1
+vendid=0x14e4
+xtalfreq=37400
diff -uprN batocera.linux-b34_origin/board/batocera/rockchip/rk3399/linux_patches/linux-add-tn3399_v3.patch batocera.linux-b34/board/batocera/rockchip/rk3399/linux_patches/linux-add-tn3399_v3.patch
--- batocera.linux-b34_origin/board/batocera/rockchip/rk3399/linux_patches/linux-add-tn3399_v3.patch	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/board/batocera/rockchip/rk3399/linux_patches/linux-add-tn3399_v3.patch	2022-06-15 21:58:39.000000000 +0800
@@ -0,0 +1,755 @@
+diff --git a/arch/arm64/boot/dts/rockchip/Makefile b/arch/arm64/boot/dts/rockchip/Makefile
+index 1ab55a124..99cfd2b91 100644
+--- a/arch/arm64/boot/dts/rockchip/Makefile
++++ b/arch/arm64/boot/dts/rockchip/Makefile
+@@ -48,3 +48,4 @@ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-rockpro64.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-sapphire.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-sapphire-excavator.dtb
+ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399pro-rock-pi-n10.dtb
++dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-tn3399_v3.dtb
+diff --git a/arch/arm64/boot/dts/rockchip/rk3399-tn3399_v3.dts b/arch/arm64/boot/dts/rockchip/rk3399-tn3399_v3.dts
+new file mode 100755
+index 000000000..0c07594dc
+--- /dev/null
++++ b/arch/arm64/boot/dts/rockchip/rk3399-tn3399_v3.dts
+@@ -0,0 +1,740 @@
++// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
++/*
++ * Copyright (c) 2017 Fuzhou Rockchip Electronics Co., Ltd.
++ */
++
++/dts-v1/;
++
++#include <dt-bindings/input/linux-event-codes.h>
++#include <dt-bindings/pwm/pwm.h>
++#include "rk3399.dtsi"
++#include "rk3399-opp.dtsi"
++
++/ {
++	model = "TN3399_V3 Board";
++	compatible = "rockchip,tn3399_v3", "rockchip,rk3399";
++
++	aliases {
++		mmc0 = &sdmmc;
++		mmc1 = &sdhci;
++		mmc2 = &sdio0;
++	};
++
++	chosen {
++		stdout-path = "serial2:1500000n8";
++	};
++
++	clkin_gmac: external-gmac-clock {
++		compatible = "fixed-clock";
++		clock-frequency = <125000000>;
++		clock-output-names = "clkin_gmac";
++		#clock-cells = <0>;
++	};
++
++	gpio-leds {
++		compatible = "gpio-leds";
++		pinctrl-names = "default";
++		pinctrl-0 = <&sys_led_pin>;
++
++		sys_led: led-0 {
++			label = "sys_led";
++			gpios = <&gpio1 RK_PC2 GPIO_ACTIVE_LOW>;
++			linux,default-trigger = "heartbeat";
++		};
++	};
++
++	sdio_pwrseq: sdio-pwrseq {
++		compatible = "mmc-pwrseq-simple";
++		clocks = <&rk808 1>;
++		clock-names = "ext_clock";
++		pinctrl-names = "default";
++		pinctrl-0 = <&wifi_enable_h>;
++		reset-gpios = <&gpio0 RK_PB2 GPIO_ACTIVE_LOW>;
++	};
++
++	vcc12v_dcin: vcc12v-dcin {
++		compatible = "regulator-fixed";
++		regulator-name = "vcc12v_dcin";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <12000000>;
++		regulator-max-microvolt = <12000000>;
++	};
++
++	vcc5v0_sys: vcc5v0-sys {
++		compatible = "regulator-fixed";
++		regulator-name = "vcc5v0_sys";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <5000000>;
++		regulator-max-microvolt = <5000000>;
++		vin-supply = <&vcc12v_dcin>;
++	};
++
++	vcc5v0_host: vcc5v0-host-regulator {
++		compatible = "regulator-fixed";
++		enable-active-high;
++		gpio = <&gpio0 RK_PA5 GPIO_ACTIVE_HIGH>;
++		pinctrl-names = "default";
++		pinctrl-0 = <&vcc5v0_host_en>;
++		regulator-name = "vcc5v0_host";
++		regulator-always-on;
++		vin-supply = <&vcc5v0_sys>;
++	};
++
++	vcc3v3_sys: vcc3v3-sys {
++		compatible = "regulator-fixed";
++		regulator-name = "vcc3v3_sys";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <3300000>;
++		regulator-max-microvolt = <3300000>;
++		vin-supply = <&vcc12v_dcin>;
++	};
++
++	vcc3v0_sd: vcc3v0-sd {
++		compatible = "regulator-fixed";
++		enable-active-high;
++		gpio = <&gpio0 RK_PA1 GPIO_ACTIVE_HIGH>;
++		pinctrl-names = "default";
++		pinctrl-0 = <&vcc3v0_sd_en>;
++		regulator-name = "vcc3v0_sd";
++		regulator-always-on;
++		vin-supply = <&vcc3v3_sys>;
++	};
++
++	vcca1v8_s3: vcca1v8-s3 {
++		compatible = "regulator-fixed";
++		regulator-name = "vcca1v8_s3";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <1800000>;
++		regulator-max-microvolt = <1800000>;
++		vin-supply = <&vcc3v3_sys>;
++	};
++
++	vcc_0v9: vcc-0v9 {
++		compatible = "regulator-fixed";
++		regulator-name = "vcc_0v9";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <900000>;
++		regulator-max-microvolt = <900000>;
++		vin-supply = <&vcc3v3_sys>;
++	};
++
++	vcc_lan: vcc-lan {
++		compatible = "regulator-fixed";
++		regulator-name = "vcc_lan";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <3300000>;
++		regulator-max-microvolt = <3300000>;
++		vin-supply = <&vcc3v3_sys>;
++	};
++
++	vdd_log: vdd-log {
++		compatible = "pwm-regulator";
++		pwms = <&pwm2 0 25000 1>;
++		regulator-name = "vdd_log";
++		regulator-always-on;
++		regulator-boot-on;
++		regulator-min-microvolt = <800000>;
++		regulator-max-microvolt = <1400000>;
++		vin-supply = <&vcc3v3_sys>;
++	};
++};
++
++&cpu_l0 {
++	cpu-supply = <&vdd_cpu_l>;
++};
++
++&cpu_l1 {
++	cpu-supply = <&vdd_cpu_l>;
++};
++
++&cpu_l2 {
++	cpu-supply = <&vdd_cpu_l>;
++};
++
++&cpu_l3 {
++	cpu-supply = <&vdd_cpu_l>;
++};
++
++&cpu_b0 {
++	cpu-supply = <&vdd_cpu_b>;
++};
++
++&cpu_b1 {
++	cpu-supply = <&vdd_cpu_b>;
++};
++
++&emmc_phy {
++	status = "okay";
++};
++
++&gmac {
++	assigned-clocks = <&cru SCLK_RMII_SRC>;
++	assigned-clock-parents = <&clkin_gmac>;
++	clock_in_out = "input";
++	phy-supply = <&vcc_lan>;
++	phy-mode = "rgmii";
++	pinctrl-names = "default";
++	pinctrl-0 = <&rgmii_pins>;
++	pinctrl-1 = <&rgmii_sleep_pins>;
++	snps,reset-gpio = <&gpio3 RK_PB7 GPIO_ACTIVE_LOW>;
++	snps,reset-active-low;
++	snps,reset-delays-us = <0 10000 50000>;
++	tx_delay = <0x28>;
++	rx_delay = <0x11>;
++	status = "okay";
++};
++
++&gpu {
++	status = "okay";
++	mali-supply = <&vdd_gpu>;
++};
++
++&hdmi {
++	ddc-i2c-bus = <&i2c3>;
++	pinctrl-names = "default";
++	pinctrl-0 = <&hdmi_cec>;
++	status = "okay";
++};
++
++&hdmi_sound {
++	status = "okay";
++};
++
++&i2c0 {
++	clock-frequency = <400000>;
++	i2c-scl-rising-time-ns = <168>;
++	i2c-scl-falling-time-ns = <4>;
++	status = "okay";
++
++	rk808: pmic@1b {
++		compatible = "rockchip,rk808";
++		reg = <0x1b>;
++		interrupt-parent = <&gpio1>;
++		interrupts = <21 IRQ_TYPE_LEVEL_LOW>;
++		#clock-cells = <1>;
++		clock-output-names = "xin32k", "rk808-clkout2";
++		pinctrl-names = "default";
++		pinctrl-0 = <&pmic_int_l>;
++		rockchip,system-power-controller;
++		wakeup-source;
++
++		vcc1-supply = <&vcc3v3_sys>;
++		vcc2-supply = <&vcc3v3_sys>;
++		vcc3-supply = <&vcc3v3_sys>;
++		vcc4-supply = <&vcc3v3_sys>;
++		vcc6-supply = <&vcc3v3_sys>;
++		vcc7-supply = <&vcc3v3_sys>;
++		vcc8-supply = <&vcc3v3_sys>;
++		vcc9-supply = <&vcc3v3_sys>;
++		vcc10-supply = <&vcc3v3_sys>;
++		vcc11-supply = <&vcc3v3_sys>;
++		vcc12-supply = <&vcc3v3_sys>;
++		vddio-supply = <&vcc_1v8>;
++
++		regulators {
++			vdd_center: DCDC_REG1 {
++				regulator-name = "vdd_center";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <750000>;
++				regulator-max-microvolt = <1350000>;
++				regulator-ramp-delay = <6001>;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++
++			vdd_cpu_l: DCDC_REG2 {
++				regulator-name = "vdd_cpu_l";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <750000>;
++				regulator-max-microvolt = <1350000>;
++				regulator-ramp-delay = <6001>;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++
++			vcc_ddr: DCDC_REG3 {
++				regulator-name = "vcc_ddr";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-state-mem {
++					regulator-on-in-suspend;
++				};
++			};
++
++			vcc_1v8: DCDC_REG4 {
++				regulator-name = "vcc_1v8";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <1800000>;
++				regulator-max-microvolt = <1800000>;
++				regulator-state-mem {
++					regulator-on-in-suspend;
++					regulator-suspend-microvolt = <1800000>;
++				};
++			};
++
++			vcc1v8_dvp: LDO_REG1 {
++				regulator-name = "vcc1v8_dvp";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <1800000>;
++				regulator-max-microvolt = <1800000>;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++					regulator-suspend-microvolt = <1800000>;
++				};
++			};
++
++			vcc3v0_touch: LDO_REG2 {
++				regulator-name = "vcc3v0_touch";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <3000000>;
++				regulator-max-microvolt = <3000000>;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++
++			vcc1v8_pmupll: LDO_REG3 {
++				regulator-name = "vcc1v8_pmupll";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <1800000>;
++				regulator-max-microvolt = <1800000>;
++				regulator-state-mem {
++					regulator-on-in-suspend;
++					regulator-suspend-microvolt = <1800000>;
++				};
++			};
++
++			vcc_sdio: LDO_REG4 {
++				regulator-name = "vcc_sdio";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <3000000>;
++				regulator-max-microvolt = <3000000>;
++				regulator-state-mem {
++					regulator-on-in-suspend;
++					regulator-suspend-microvolt = <3000000>;
++				};
++			};
++
++			vcca3v0_codec: LDO_REG5 {
++				regulator-name = "vcca3v0_codec";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <3000000>;
++				regulator-max-microvolt = <3000000>;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++
++			vcc_1v5: LDO_REG6 {
++				regulator-name = "vcc_1v5";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <1500000>;
++				regulator-max-microvolt = <1500000>;
++				regulator-state-mem {
++					regulator-on-in-suspend;
++					regulator-suspend-microvolt = <1500000>;
++				};
++			};
++
++			vcca1v8_codec: LDO_REG7 {
++				regulator-name = "vcca1v8_codec";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <1800000>;
++				regulator-max-microvolt = <1800000>;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++
++			vcc_3v0: LDO_REG8 {
++				regulator-name = "vcc_3v0";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-min-microvolt = <3000000>;
++				regulator-max-microvolt = <3000000>;
++				regulator-state-mem {
++					regulator-on-in-suspend;
++					regulator-suspend-microvolt = <3000000>;
++				};
++			};
++
++			vcc3v3_s3: SWITCH_REG1 {
++				regulator-name = "vcc3v3_s3";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++
++			vcc3v3_s0: SWITCH_REG2 {
++				regulator-name = "vcc3v3_s0";
++				regulator-always-on;
++				regulator-boot-on;
++				regulator-state-mem {
++					regulator-off-in-suspend;
++				};
++			};
++		};
++	};
++
++	vdd_cpu_b: regulator@40 {
++		compatible = "silergy,syr827";
++		reg = <0x40>;
++		fcs,suspend-voltage-selector = <1>;
++		pinctrl-names = "default";
++		pinctrl-0 = <&vsel1_pin>;
++		regulator-name = "vdd_cpu_b";
++		regulator-min-microvolt = <712500>;
++		regulator-max-microvolt = <1500000>;
++		regulator-ramp-delay = <1000>;
++		regulator-always-on;
++		regulator-boot-on;
++		vin-supply = <&vcc3v3_sys>;
++
++		regulator-state-mem {
++			regulator-off-in-suspend;
++		};
++	};
++
++	vdd_gpu: regulator@41 {
++		compatible = "silergy,syr828";
++		reg = <0x41>;
++		fcs,suspend-voltage-selector = <1>;
++		pinctrl-names = "default";
++		pinctrl-0 = <&vsel2_pin>;
++		regulator-name = "vdd_gpu";
++		regulator-min-microvolt = <712500>;
++		regulator-max-microvolt = <1500000>;
++		regulator-ramp-delay = <1000>;
++		regulator-always-on;
++		regulator-boot-on;
++		vin-supply = <&vcc3v3_sys>;
++
++		regulator-state-mem {
++			regulator-off-in-suspend;
++		};
++	};
++};
++
++&i2c1 {
++	status = "okay";
++};
++
++&i2c2 {
++	status = "okay";
++};
++
++&i2c3 {
++	status = "okay";
++};
++
++&i2c4 {
++	status = "okay";
++};
++
++&i2c7 {
++	status = "okay";
++};
++
++&i2s0 {
++	rockchip,playback-channels = <8>;
++	rockchip,capture-channels = <8>;
++	status = "okay";
++};
++
++&i2s1 {
++	rockchip,playback-channels = <2>;
++	rockchip,capture-channels = <2>;
++	status = "okay";
++};
++
++&i2s2 {
++	status = "okay";
++};
++
++&io_domains {
++	status = "okay";
++
++	bt656-supply = <&vcc1v8_dvp>;
++	audio-supply = <&vcca1v8_codec>;
++	sdmmc-supply = <&vcc_sdio>;
++	gpio1830-supply = <&vcc_3v0>;
++};
++
++&pmu_io_domains {
++	status = "okay";
++	pmu1830-supply = <&vcc_1v8>;
++};
++
++&pinctrl {
++	leds {
++		sys_led_pin: sys-led-pin {
++			rockchip,pins = <1 RK_PC2 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++	};
++
++	bt {
++		bt_enable_h: bt-enable-h {
++			rockchip,pins = <0 RK_PB1 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++
++		bt_host_wake_l: bt-host-wake-l {
++			rockchip,pins = <0 RK_PA4 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++
++		bt_wake_l: bt-wake-l {
++			rockchip,pins = <2 RK_PD2 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++	};
++
++	wifi {
++		wifi_enable_h: wifi-enable-h {
++			rockchip,pins = <0 RK_PB2 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++
++		wifi_host_wake_l: wifi-host-wake-l {
++			rockchip,pins = <0 RK_PA3 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++	};
++
++	usb2 {
++		vcc5v0_host_en: vcc5v0-host-en {
++			rockchip,pins = <0 RK_PA5 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++	};
++
++	sdmmc {
++		vcc3v0_sd_en: vcc3v0-sd-en {
++			rockchip,pins =<0 RK_PA1 RK_FUNC_GPIO &pcfg_pull_none>;
++		};
++	};
++
++	pmic {
++		pmic_int_l: pmic-int-l {
++			rockchip,pins = <1 RK_PC5 RK_FUNC_GPIO &pcfg_pull_up>;
++		};
++
++		vsel1_pin: vsel1-pin {
++			rockchip,pins = <1 RK_PC1 RK_FUNC_GPIO &pcfg_pull_down>;
++		};
++
++		vsel2_pin: vsel2-pin {
++			rockchip,pins = <1 RK_PB6 RK_FUNC_GPIO &pcfg_pull_down>;
++		};
++	};
++
++	gmac {
++		rgmii_sleep_pins: rgmii-sleep-pins {
++			rockchip,pins =<3 RK_PB7 RK_FUNC_GPIO &pcfg_output_low>;
++		};
++	};
++};
++
++&pwm0 {
++	status = "okay";
++};
++
++&pwm2 {
++	status = "okay";
++};
++
++&pwm3 {
++	status = "okay";
++};
++
++&saradc {
++	status = "okay";
++	vref-supply = <&vcca1v8_s3>;
++};
++
++&sdio0 {
++	status = "okay";
++
++	#address-cells = <1>;
++	#size-cells = <0>;
++	bus-width = <4>;
++	clock-frequency = <50000000>;
++	cap-sdio-irq;
++	cap-sd-highspeed;
++	keep-power-in-suspend;
++	mmc-pwrseq = <&sdio_pwrseq>;
++	non-removable;
++	pinctrl-names = "default";
++	pinctrl-0 = <&sdio0_bus4 &sdio0_cmd &sdio0_clk>;
++	sd-uhs-sdr104;
++
++	brcmf: wifi@1 {
++		compatible = "brcm,bcm43455-fmac";
++		reg = <1>;
++		interrupt-parent = <&gpio0>;
++		interrupts = <RK_PA3 GPIO_ACTIVE_HIGH>;
++		interrupt-names = "host-wake";
++		pinctrl-names = "default";
++		pinctrl-0 = <&wifi_host_wake_l>;
++	};
++};
++
++&sdmmc {
++	bus-width = <4>;
++	cap-mmc-highspeed;
++	cap-sd-highspeed;
++	cd-gpios = <&gpio0 RK_PA7 GPIO_ACTIVE_LOW>;
++	disable-wp;
++	max-frequency = <150000000>;
++	pinctrl-names = "default";
++	pinctrl-0 = <&sdmmc_clk &sdmmc_cd &sdmmc_cmd &sdmmc_bus4>;
++	vmmc-supply = <&vcc3v0_sd>;
++	vqmmc-supply = <&vcc_sdio>;
++	status = "okay";
++};
++
++&sdhci {
++	bus-width = <8>;
++	mmc-hs400-1_8v;
++	mmc-hs400-enhanced-strobe;
++	non-removable;
++	status = "okay";
++};
++
++&spi1 {
++	status = "okay";
++};
++
++&tcphy0 {
++	status = "okay";
++};
++
++&tcphy1 {
++	status = "okay";
++};
++
++&tsadc {
++	status = "okay";
++
++	/* tshut mode 0:CRU 1:GPIO */
++	rockchip,hw-tshut-mode = <1>;
++	/* tshut polarity 0:LOW 1:HIGH */
++	rockchip,hw-tshut-polarity = <1>;
++};
++
++&u2phy0 {
++	status = "okay";
++
++	u2phy0_otg: otg-port {
++		status = "disabled";
++	};
++
++	u2phy0_host: host-port {
++		phy-supply = <&vcc5v0_host>;
++		status = "okay";
++	};
++};
++
++&u2phy1 {
++	status = "okay";
++
++	u2phy1_otg: otg-port {
++		status = "disabled";
++	};
++
++	u2phy1_host: host-port {
++		phy-supply = <&vcc5v0_host>;
++		status = "okay";
++	};
++};
++
++&uart0 {
++	pinctrl-names = "default";
++	pinctrl-0 = <&uart0_xfer &uart0_cts &uart0_rts>;
++
++	status = "okay";
++
++	bluetooth {
++		compatible = "brcm,bcm4345c5";
++		clocks = <&rk808 1>;
++		clock-names = "ext_clock";
++		device-wakeup-gpios = <&gpio2 RK_PD2 GPIO_ACTIVE_HIGH>;
++		host-wakeup-gpios = <&gpio0 RK_PA4 GPIO_ACTIVE_HIGH>;
++		shutdown-gpios = <&gpio0 RK_PB1 GPIO_ACTIVE_HIGH>;
++		max-speed = <1500000>;
++		pinctrl-names = "default";
++		pinctrl-0 = <&bt_host_wake_l &bt_wake_l &bt_enable_h>;
++	};
++};
++
++&uart2 {
++	status = "okay";
++};
++
++&usb_host0_ehci {
++	status = "okay";
++};
++
++&usb_host0_ohci {
++	status = "okay";
++};
++
++&usb_host1_ehci {
++	status = "okay";
++};
++
++&usb_host1_ohci {
++	status = "okay";
++};
++
++&usbdrd3_0 {
++	status = "okay";
++};
++
++&usbdrd_dwc3_0 {
++	status = "okay";
++	dr_mode = "host";
++};
++
++&usbdrd3_1 {
++	status = "okay";
++};
++
++&usbdrd_dwc3_1 {
++	status = "okay";
++	dr_mode = "host";
++};
++
++&vopb {
++	status = "okay";
++};
++
++&vopb_mmu {
++	status = "okay";
++};
++
++&vopl {
++	status = "okay";
++};
++
++&vopl_mmu {
++	status = "okay";
++};
++
++&iep_mmu {
++	status = "okay";
++};
diff -uprN batocera.linux-b34_origin/board/batocera/rockchip/rk3399/tn3399_v3/boot/extlinux.conf batocera.linux-b34/board/batocera/rockchip/rk3399/tn3399_v3/boot/extlinux.conf
--- batocera.linux-b34_origin/board/batocera/rockchip/rk3399/tn3399_v3/boot/extlinux.conf	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/board/batocera/rockchip/rk3399/tn3399_v3/boot/extlinux.conf	2022-06-15 22:01:29.000000000 +0800
@@ -0,0 +1,4 @@
+LABEL batocera.linux
+  LINUX /boot/linux
+  FDT /boot/rk3399-tn3399_v3.dtb
+  APPEND initrd=/boot/initrd.gz label=BATOCERA rootwait quiet loglevel=0 console=tty3 console=ttyS2,1500000n8
diff -uprN batocera.linux-b34_origin/board/batocera/rockchip/rk3399/tn3399_v3/create-boot-script.sh batocera.linux-b34/board/batocera/rockchip/rk3399/tn3399_v3/create-boot-script.sh
--- batocera.linux-b34_origin/board/batocera/rockchip/rk3399/tn3399_v3/create-boot-script.sh	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/board/batocera/rockchip/rk3399/tn3399_v3/create-boot-script.sh	2022-06-15 22:02:05.000000000 +0800
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+# HOST_DIR = host dir
+# BOARD_DIR = board specific dir
+# BUILD_DIR = base dir/build
+# BINARIES_DIR = images dir
+# TARGET_DIR = target dir
+# BATOCERA_BINARIES_DIR = batocera binaries sub directory
+
+HOST_DIR=$1
+BOARD_DIR=$2
+BUILD_DIR=$3
+BINARIES_DIR=$4
+TARGET_DIR=$5
+BATOCERA_BINARIES_DIR=$6
+
+mkdir -p "${BATOCERA_BINARIES_DIR}/boot/boot"     || exit 1
+mkdir -p "${BATOCERA_BINARIES_DIR}/boot/extlinux" || exit 1
+
+cp "${BINARIES_DIR}/Image"                 "${BATOCERA_BINARIES_DIR}/boot/boot/linux"                || exit 1
+cp "${BINARIES_DIR}/initrd.gz"             "${BATOCERA_BINARIES_DIR}/boot/boot/initrd.gz"            || exit 1
+cp "${BINARIES_DIR}/rootfs.squashfs"       "${BATOCERA_BINARIES_DIR}/boot/boot/batocera.update"      || exit 1
+
+cp "${BINARIES_DIR}/rk3399-tn3399_v3.dtb"  "${BATOCERA_BINARIES_DIR}/boot/boot/"     || exit 1
+cp "${BOARD_DIR}/boot/extlinux.conf"       "${BATOCERA_BINARIES_DIR}/boot/extlinux/" || exit 1
+
+exit 0
diff -uprN batocera.linux-b34_origin/board/batocera/rockchip/rk3399/tn3399_v3/genimage.cfg batocera.linux-b34/board/batocera/rockchip/rk3399/tn3399_v3/genimage.cfg
--- batocera.linux-b34_origin/board/batocera/rockchip/rk3399/tn3399_v3/genimage.cfg	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/board/batocera/rockchip/rk3399/tn3399_v3/genimage.cfg	2022-06-16 09:44:56.000000000 +0800
@@ -0,0 +1,41 @@
+image boot.vfat {
+	vfat {
+		extraargs = "-F 32 -n BATOCERA"
+		@files
+	}
+	size = 4G
+}
+
+image userdata.ext4 {
+	ext4 {
+		label = "SHARE"
+		use-mke2fs = "true"
+		extraargs = "-m 0"
+	}
+	size = "256M"
+	# include files from TARGET_DIR/userdata
+	mountpoint = "/userdata"
+}
+
+image batocera.img {
+	hdimage {
+		align = "1M"
+	}
+
+	partition part_uboot {
+		in-partition-table = "no"
+		image = "../../tn3399_v3/u-boot-tn3399_v3.bin"
+		offset = 32K
+	}
+
+	partition vfat {
+		partition-type = 0xC
+		image = "boot.vfat"
+		offset = 16M
+	}
+
+	partition userdata {
+		partition-type = 0x83
+		image = "userdata.ext4"
+	}
+}
diff -uprN batocera.linux-b34_origin/Config.in batocera.linux-b34/Config.in
--- batocera.linux-b34_origin/Config.in	2022-05-26 23:56:31.000000000 +0800
+++ batocera.linux-b34/Config.in	2022-06-16 09:38:00.000000000 +0800
@@ -32,6 +32,7 @@ menu "System"
     source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-odroid-c2/Config.in"
     source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-s905-tvbox/Config.in"
     source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-rk3399/Config.in"
+    source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-tn3399_v3/Config.in"
     source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-ps5000/Config.in"
     source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-ps7000/Config.in"
     source "$BR2_EXTERNAL_BATOCERA_PATH/package/batocera/boot/uboot-powkiddy-a13/Config.in"
diff -uprN batocera.linux-b34_origin/configs/batocera-rk3399_defconfig batocera.linux-b34/configs/batocera-rk3399_defconfig
--- batocera.linux-b34_origin/configs/batocera-rk3399_defconfig	2022-05-26 23:56:31.000000000 +0800
+++ batocera.linux-b34/configs/batocera-rk3399_defconfig	2022-06-16 09:39:51.000000000 +0800
@@ -43,7 +43,7 @@ BR2_LINUX_KERNEL_USE_CUSTOM_CONFIG=y
 BR2_LINUX_KERNEL_CUSTOM_CONFIG_FILE="$(BR2_EXTERNAL_BATOCERA_PATH)/board/batocera/rockchip/rk3399/linux-rk3399-defconfig.config"
 BR2_LINUX_KERNEL_CONFIG_FRAGMENT_FILES="$(BR2_EXTERNAL_BATOCERA_PATH)/board/batocera/rockchip/rk3399/linux-defconfig-fragment.config"
 BR2_LINUX_KERNEL_DTS_SUPPORT=y
-BR2_LINUX_KERNEL_INTREE_DTS_NAME="rockchip/rk3399-rockpro64 rockchip/rk3399-rock960 rockchip/rk3399-hugsun-x99"
+BR2_LINUX_KERNEL_INTREE_DTS_NAME="rockchip/rk3399-rockpro64 rockchip/rk3399-rock960 rockchip/rk3399-hugsun-x99 rockchip/rk3399-tn3399_v3"
 BR2_LINUX_KERNEL_NEEDS_HOST_OPENSSL=y
 BR2_LINUX_KERNEL_IMAGE_TARGET_CUSTOM=y
 BR2_LINUX_KERNEL_IMAGE_TARGET_NAME="Image"
@@ -52,6 +52,7 @@ BR2_LINUX_KERNEL_IMAGE_TARGET_NAME="Imag
 
 # Bootloader (prebuilt)
 BR2_PACKAGE_UBOOT_RK3399=y
+BR2_PACKAGE_UBOOT_TN3399_V3=y
 
 # System
 BR2_PACKAGE_BUSYBOX_CONFIG_FRAGMENT_FILES="$(BR2_EXTERNAL_BATOCERA_PATH)/board/batocera/busybox.custom.config"
diff -uprN batocera.linux-b34_origin/package/batocera/boot/uboot-tn3399_v3/Config.in batocera.linux-b34/package/batocera/boot/uboot-tn3399_v3/Config.in
--- batocera.linux-b34_origin/package/batocera/boot/uboot-tn3399_v3/Config.in	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/package/batocera/boot/uboot-tn3399_v3/Config.in	2022-06-16 09:41:46.000000000 +0800
@@ -0,0 +1,4 @@
+config BR2_PACKAGE_UBOOT_TN3399_V3
+        bool "uboot files for tn3399_v3"
+        help
+          uboot files for tn3399_v3
diff -uprN batocera.linux-b34_origin/package/batocera/boot/uboot-tn3399_v3/uboot-tn3399_v3.mk batocera.linux-b34/package/batocera/boot/uboot-tn3399_v3/uboot-tn3399_v3.mk
--- batocera.linux-b34_origin/package/batocera/boot/uboot-tn3399_v3/uboot-tn3399_v3.mk	1970-01-01 08:00:00.000000000 +0800
+++ batocera.linux-b34/package/batocera/boot/uboot-tn3399_v3/uboot-tn3399_v3.mk	2022-06-16 11:47:07.000000000 +0800
@@ -0,0 +1,15 @@
+################################################################################
+#
+# uboot files for TN3399_V3
+#
+################################################################################
+UBOOT_TN3399_V3_VERSION = 326cc3a3f5994340037394eef20b39d24b1cf93e
+UBOOT_TN3399_V3_SITE = https://github.com/retro98boy/u-boot-bin-tn3399_v3.git
+UBOOT_TN3399_V3_SITE_METHOD=git
+
+define UBOOT_TN3399_V3_INSTALL_TARGET_CMDS
+	mkdir -p   $(BINARIES_DIR)/tn3399_v3/
+	cp $(@D)/* $(BINARIES_DIR)/tn3399_v3/
+endef
+
+$(eval $(generic-package))
diff -uprN batocera.linux-b34_origin/package/batocera/core/batocera-system/Config.in batocera.linux-b34/package/batocera/core/batocera-system/Config.in
--- batocera.linux-b34_origin/package/batocera/core/batocera-system/Config.in	2022-05-26 23:56:31.000000000 +0800
+++ batocera.linux-b34/package/batocera/core/batocera-system/Config.in	2022-06-16 09:46:27.000000000 +0800
@@ -40,7 +40,7 @@ endif
 # add platforms that support GLES3 here
 # the GLES3 backend maps to OpenGL 3.3 on desktop, OpenGL ES 3.0 on mobile and WebGL 2.0 on the web.
 if BR2_PACKAGE_BATOCERA_TARGET_X86_ANY || BR2_PACKAGE_BATOCERA_TARGET_RPI4 || BR2_PACKAGE_BATOCERA_TARGET_RK3399 || BR2_PACKAGE_BATOCERA_TARGET_RG552 \
-    || BR2_PACKAGE_BATOCERA_TARGET_RK3326 || BR2_PACKAGE_BATOCERA_TARGET_XU4 || BR2_PACKAGE_BATOCERA_TARGET_ORANGEPI_ZERO2 
+    || BR2_PACKAGE_BATOCERA_TARGET_RK3326 || BR2_PACKAGE_BATOCERA_TARGET_XU4 || BR2_PACKAGE_BATOCERA_TARGET_ORANGEPI_ZERO2
     config BR2_PACKAGE_BATOCERA_GLES3
 	bool "target gles3 support"
 	default y
@@ -166,7 +166,7 @@ config BR2_PACKAGE_BATOCERA_SYSTEM
 	select BR2_PACKAGE_OPENVPN                              # VPN support
 	select BR2_PACKAGE_USB_MODESWITCH                       # USB modeswitch support
 	select BR2_PACKAGE_UMTOOL								# command line tool for configuring Ultimarc arcade hardware
-	
+
 	# Deprecated, which what library should we replace ?
 	select BR2_PACKAGE_PARTED                               # partition management (for the first boot)
 	select BR2_PACKAGE_USBMOUNT                             # usb key/sd card mounter
@@ -248,7 +248,7 @@ config BR2_PACKAGE_BATOCERA_SYSTEM
                                                 !BR2_PACKAGE_BATOCERA_TARGET_RK3326         && \
                                                 !BR2_PACKAGE_BATOCERA_TARGET_RK3128         && \
 												!BR2_PACKAGE_BATOCERA_TARGET_S922X
-		
+
 	select BR2_PACKAGE_RTL8812AU        if BR2_PACKAGE_BATOCERA_TARGET_S922X
 	select BR2_PACKAGE_RTL8821CU        if BR2_PACKAGE_BATOCERA_TARGET_S922X
 
@@ -353,7 +353,7 @@ config BR2_TARGET_BATOCERA_IMAGES_DEFINI
 	default "raspberrypi/rpizero2"                                  if BR2_PACKAGE_BATOCERA_TARGET_RPIZERO2
     default "rockchip/rk3326/rk3326 rockchip/rk3326/gameforce"      if BR2_PACKAGE_BATOCERA_TARGET_RK3326
 	default "rockchip/rk3288/tinkerboard rockchip/rk3288/miqi"      if BR2_PACKAGE_BATOCERA_TARGET_RK3288
-	default "rockchip/rk3399/rock960 rockchip/rk3399/rockpro64"     if BR2_PACKAGE_BATOCERA_TARGET_RK3399
+	default "rockchip/rk3399/rock960 rockchip/rk3399/rockpro64 rockchip/rk3399/tn3399_v3"     if BR2_PACKAGE_BATOCERA_TARGET_RK3399
 	default "rockchip/rg552"                                        if BR2_PACKAGE_BATOCERA_TARGET_RG552
 	default "rockchip/rk3128/ps5000 rockchip/rk3128/ps7000 rockchip/rk3128/powkiddy_a13"     if BR2_PACKAGE_BATOCERA_TARGET_RK3128
 	default "tritium-h5"                                            if BR2_PACKAGE_BATOCERA_TARGET_TRITIUM_H5
@@ -1099,7 +1099,7 @@ config BR2_PACKAGE_BATOCERA_KODI19
 	select BR2_PACKAGE_FFMPEG_AVRESAMPLE    # audio resampler
 	select BR2_PACKAGE_LIBCDIO_PARANOIA     # cd-audio
 	select BR2_PACKAGE_LIBOPENH264          # openh264
-    
+
     # AV1 support for x86_64
     select BR2_PACKAGE_DAV1D           if BR2_PACKAGE_BATOCERA_TARGET_X86_64       # AV1
 
